Cmake cpp14

wsnipex opened the pull request:
<!--- Provide a general summary of your change in the Title above -->
rebased https://github.com/xbmc/xbmc/pull/10864 and added depends check

note: the test commit needs to be dropped before merge

notspiff commented on cmake/scripts/common/ArchSetup.cmake:
```diff
@@ -98,11 +98,11 @@ message(STATUS "Cross-Compiling: ${CMAKE_CROSSCOMPILING}")
+check_type(char16_t HAVE_CHAR16_T)
+check_type(char32_t HAVE_CHAR32_T)
+check_type(std::u16string HAVE_STD__U16_STRING)
+check_type(std::u32string HAVE_STD__U32_STRING)
```
these are guaranteed by c++14 (actually, by 11) and since we require the standard, the probes are useless no ?

wsnipex replied:
good point. iirc we had them even before c++11. I can remove them

notspiff replied:
yes because we initially built under c++0x. you have to change some code that checks the defines as well

notspiff replied:
actually the code using it is nowhere to be found so just removing the probes suffices.

Paxxi commented:
Just curious, is it Ubuntu or Android holding us back here with gcc versions?

Going up to 17 is obviously an issue but depending on our minimum gcc version std1z could have been an option, gaining some library features

wsnipex commented:
on the ubuntu side there is only 14.04 that is a problem, it has only gcc 4.8. Since it's showing it's age on the kernel and mesa side, it's time to drop support anyway(I already did for nightlies).
Then there is Rbpi, which has gcc 4.9(min required for c++14) since 2 days ago.
All other platforms have newer toolchains.

https://gcc.gnu.org/projects/cxx-status.html shows that std1z is currently only well supported in gcc 7.

Paxxi commented:
yeah, was thinking more about standard library bits like filesystem etc but it doesn't look much better there so it's a moot point. Will have to be patient 😄

wsnipex commented:
jenkins build this please

Rechi commented on cmake/scripts/common/ArchSetup.cmake:
```diff
@@ -15,20 +15,20 @@ include(CheckCXXSourceCompiles)
-                             {
-                               ${type} s;
-                             }" ${var})
+macro(check_type type var)
```
macro is unused

wsnipex commented:
jenkins build this please

wsnipex commented:
good to go?

Rechi commented:
There are still some cxx11 and c++11 settings in CMakeLists.txt and in the cmake folder which should be either update to c++14 or removed if not needed any more.

wsnipex commented:
jenkins build this please

wsnipex commented:
jenkins build this please

wsnipex commented:
know that tests should be reliable:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

wsnipex commented:
dumdedeedum
jenkins build this please

MartijnKaijser commented:
Jenkins build this please

MartijnKaijser commented:
Jenkins build this please
Good for merge once build successfully?

Rechi commented:
no, Windows crash in CVariant with this PR

lrusak commented:
@wsnipex what's holding us back here? Can we have this in for v19?

wsnipex commented:
there where (suspected) windows crashes. No idea if that is still the case, but if it is, I won't be able to fix it.

Rechi commented:
Seems like https://forum.kodi.tv/showthread.php?tid=344300 is the same crash on Linux, I experienced on Windows. So most likely not a Windows toolchain bug, but a bug in Kodi code.

fritsch commented:
I also was interested but could not access the provided crashlog, do you still have it?

Rechi commented:
no

a1rwulf commented:
The user has reposted the crashlog, are we sure this crash is caused by the cpp14 switch.
I'm pretty sure I've seen this already with kodi 18 as well as current master:
```
Thread 1 (Thread 0x7f41435de700 (LWP 4480)):
#0  0x00007f41f2e830e5 in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000055e978e7d793 in CVariant::operator=(CVariant const&) ()
#2  0x000055e9793ce566 in JSONRPC::JSONSchemaTypeDefinition::Set(std::shared_ptr<JSONRPC::JSONSchemaTypeDefinition>) ()
#3  0x000055e9793ee8ff in JSONRPC::JSONSchemaTypeDefinition::Check(CVariant const&, CVariant&, CVariant&) ()
#4  0x000055e9793f0a73 in JSONRPC::JsonRpcMethod::checkParameter(CVariant const&, std::shared_ptr<JSONRPC::JSONSchemaTypeDefinition>, unsigned int, CVariant&, unsigned int&, CVariant&) ()
#5  0x000055e9793f0f39 in JSONRPC::JsonRpcMethod::Check(CVariant const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, bool, JSONRPC::JSONRPC_STATUS (*&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, CVariant const&, CVariant&), CVariant&) const ()
#6  0x000055e9793f14e8 in JSONRPC::CJSONRPC::HandleMethodCall(CVariant const&, CVariant&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#7  0x000055e9793f1de5 in JSONRPC::CJSONRPC::MethodCall(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#8  0x000055e978aef776 in CHTTPJsonRpcHandler::HandleRequest() ()
#9  0x000055e9793b8f07 in CWebServer::HandleRequest(std::shared_ptr<IHTTPRequestHandler> const&) ()
#10 0x000055e97939215e in CWebServer::HandlePartialRequest(MHD_Connection*, CWebServer::ConnectionHandler*, HTTPRequest const&, char const*, unsigned long*, void**) ()
#11 0x000055e979387b6e in CWebServer::AnswerToConnection(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) ()
#12 0x00007f41f9a205f5 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#13 0x00007f41f9a221a8 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#14 0x00007f41f9a24016 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#15 0x00007f41f9a27558 in ?? () from /usr/lib/x86_64-linux-gnu/libmicrohttpd.so.12
#16 0x00007f41fb1246db in start_thread (arg=0x7f41435de700) at pthread_create.c:463
#17 0x00007f41f3dbc88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

a1rwulf commented:
For comparison, this is a crash I see sometimes right after startup when a json request is done.
Same place but looks like it crashes differently in the stdlib:
```
Thread 1 (Thread 0x7fe9a5e6a700 (LWP 901)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fe9ea05f524 in __GI_abort () at abort.c:79
#2  0x00007fe9e9f4e8b2 in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/libstdc++.so.6
#3  0x00007fe9e9f4c986 in ?? () from /usr/lib/libstdc++.so.6
#4  0x00007fe9e9f4c9c1 in std::terminate() () from /usr/lib/libstdc++.so.6
#5  0x00007fe9e9f4cbf3 in __cxa_throw () from /usr/lib/libstdc++.so.6
#6  0x00007fe9e9f4d0bc in operator new(unsigned long) () from /usr/lib/libstdc++.so.6
#7  0x00007fe9e9f8afa9 in std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&) () from /usr/lib/libstdc++.so.6
#8  0x00007fe9e9f8be1b in std::string::_Rep::_M_clone(std::allocator<char> const&, unsigned long) () from /usr/lib/libstdc++.so.6
#9  0x00007fe9e9f8c4fc in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&) () from /usr/lib/libstdc++.so.6
#10 0x0000000000f77926 in CVariant::operator=(CVariant const&) ()
#11 0x0000000000bdeda1 in ?? ()
#12 0x0000000000becfad in ?? ()
#13 0x0000000000beda09 in JSONRPC::JSONSchemaTypeDefinition::Set(std::shared_ptr<JSONRPC::JSONSchemaTypeDefinition>) ()
#14 0x0000000000c071fa in JSONRPC::JSONSchemaTypeDefinition::Check(CVariant const&, CVariant&, CVariant&) ()
#15 0x0000000000c078df in JSONRPC::JSONSchemaTypeDefinition::Check(CVariant const&, CVariant&, CVariant&) ()
#16 0x0000000000c09392 in JSONRPC::JsonRpcMethod::checkParameter(CVariant const&, std::shared_ptr<JSONRPC::JSONSchemaTypeDefinition>, unsigned int, CVariant&, unsigned int&, CVariant&) ()
#17 0x0000000000c096ee in JSONRPC::JsonRpcMethod::Check(CVariant const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, bool, JSONRPC::JSONRPC_STATUS (*&)(std::string const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, CVariant const&, CVariant&), CVariant&) const ()
#18 0x0000000000c09882 in JSONRPC::CJSONServiceDescription::CheckCall(char const*, CVariant const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, bool, JSONRPC::JSONRPC_STATUS (*&)(std::string const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*, CVariant const&, CVariant&), CVariant&) ()
#19 0x0000000000c09d91 in JSONRPC::CJSONRPC::HandleMethodCall(CVariant const&, CVariant&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#20 0x0000000000c0a01f in JSONRPC::CJSONRPC::MethodCall(std::string const&, JSONRPC::ITransportLayer*, JSONRPC::IClient*) ()
#21 0x00000000011f9b1d in CHTTPJsonRpcHandler::HandleRequest() ()
#22 0x0000000000cc1a8c in CWebServer::HandleRequest(std::shared_ptr<IHTTPRequestHandler> const&) ()
#23 0x0000000000cc1f9b in CWebServer::HandlePartialRequest(MHD_Connection*, CWebServer::ConnectionHandler*, HTTPRequest const&, char const*, unsigned long*, void**) ()
#24 0x0000000000cbeee1 in CWebServer::AnswerToConnection(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) ()
#25 0x0000000001233bf1 in ?? ()
#26 0x0000000001235402 in ?? ()
#27 0x0000000001235f4e in ?? ()
#28 0x00000000012384bf in ?? ()
#29 0x00007fe9ec56efa0 in start_thread (arg=<optimized out>) at pthread_create.c:486
#30 0x00007fe9ea131d5f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

a1rwulf commented:
@Rechi so is above crash still blocking this particular PR?
I think it's an unrelated issue, that seems hard to track down.
As posted above, I've experienced a similar crash with the current code as well, so it looks unrelated to the c++14 changes to me.

Paxxi commented:
I haven't looked at the code but the stack trace seems like the variant is in a bad state, possibly moved from. I know we had crashes because of that before but I thought I fixed all the cases of that 😀

a1rwulf commented:
I'll try if I can reproduce it somehow.
Right now, I can't really trigger it on my dev machine in a reliable way.

Paxxi commented:
I'm thinking that this should solve the issue. Modifying the schema definitions during a check isn't thread safe and there's no locking to protect it. https://github.com/Paxxi/xbmc/commit/164a1219622bbd7510a9091fe8e1405a1985d08a

Rechi commented:
@Paxxi this seems to fix the crash. Are you going to create a PR?

Paxxi commented:
Yeah I can do that. Since it was mentioned here I thought it could be picked into this pr

Rechi commented:
As the fix is unrelated to this PR, it should be a separate PR.

Rechi commented:
jenkins build this please

fetzerch committed:
[cmake] Enable C++14 support for core

Also remove the FindCXX11.cmake module and make use of CMake
techniques to specify the necessary compiler flags.

wsnipex committed:
[docs] update linux readmes

wsnipex committed:
[depends] check for c++14 support

Rechi commented:
jenkins build this please

AlwinEsch commented on docs/README.openSUSE.md:
```diff
@@ -72,6 +72,8 @@ sudo zypper ref
 If you get a `package not found` type of message with the below command, remove the offending package(s) from the install list and reissue the command. Take a note of the missing dependencies and, after a successful step completion, **[build the missing dependencies manually](#31-build-missing-dependencies)**.
 
+**NOTE:** Kodi requires a compiler with C++14 support, i.e. gcc >= 4.9 or clang >= 3.4
+
```
Not related to here, but must not also for Mac updated https://github.com/xbmc/xbmc/blame/master/docs/README.macOS.md#L61 ? Not so much experience with xcode, but I think Xcode 8.x can't do it.

Rechi replied:
Xcode 8.0 uses clang 3.9.0svn, https://en.wikipedia.org/wiki/Xcode#Xcode_7.0_-_11.x_(since_Free_On-Device_Development), so everything is fine

AlwinEsch commented on tools/depends/target/crossguid/CMakeLists.txt:
```diff
@@ -1,12 +1,12 @@
+cmake_minimum_required(VERSION 3.1)
```
Better change direct to 3.4 (or 3.5) to have equal to the base.

Rechi replied:
This PR is about switching to C++14, not about unifying `cmake_minimum_required`.
Version 3.1 is enough for this change.
